docs/question/server.md

```diff
@@ -107,18 +205,31 @@ The `pl.to_json` function supports keyword-only options for different types of e
 From within `server.py` functions, directories can be accessed as:
 
 ```python
-data["options"]["question_path"]                      # on-disk location of the current question directory
-data["options"]["client_files_question_path"]         # on-disk location of clientFilesQuestion/
-data["options"]["client_files_question_url"]          # URL location of clientFilesQuestion/ (only in render() function)
-data["options"]["client_files_question_dynamic_url"]  # URL location of dynamically-generated question files (only in render() function)
-data["options"]["client_files_course_path"]           # on-disk location of clientFilesCourse/
-data["options"]["client_files_course_url"]            # URL location of clientFilesCourse/ (only in render() function)
-data["options"]["server_files_course_path"]           # on-disk location of serverFilesCourse/
+# on-disk location of the current question directory
+data["options"]["question_path"]
+
+# on-disk location of clientFilesQuestion/
+data["options"]["client_files_question_path"]
+
+# URL location of clientFilesQuestion/ (only in render() function)
+data["options"]["client_files_question_url"]
+
+# URL location of dynamically-generated question files (only in render() function)
+data["options"]["client_files_question_dynamic_url"]
+
+# on-disk location of clientFilesCourse/
+data["options"]["client_files_course_path"]
+
+# URL location of clientFilesCourse/ (only in render() function)
+data["options"]["client_files_course_url"]
+
+# on-disk location of serverFilesCourse/
+data["options"]["server_files_course_path"]
 ```
 
-## Generating dynamic files
+## Generating dynamic files with `file()`
 
-You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
+You can dynamically generate file objects in `server.py`. These files never appear physically on the disk. They are generated in `file()` and returned as strings, bytes-like objects, or file-like objects. These files are generated before the question is rendered in `question.html`. A complete `question.html` and `server.py` example using a dynamically generated `fig.png` looks like:
```

Comment:
The current lifecycle diagram on `index.md` doesn't include this step. Is it worth including?

Reply:
It can't hurt!

Reply:
Added in [9b81bf8](https://github.com/PrairieLearn/PrairieLearn/pull/11830/commits/9b81bf837e7c933a426590a9586f3c3b9bf27c5b).